### PR TITLE
Optimize elliptic solver (4): Boundary conditions in MinusLaplacian preconditioner

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This is an optimization of the MinusLaplacian subdomain preconditioner. It greatly improves the effectiveness of the preconditioner when Neumann boundary conditions are used for some variables.

This PR implements the strategy described in Section III.E.1 in https://arxiv.org/abs/2111.06767. Results are shown in Fig. 15.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp [hunk 1]
```diff
@@ -79,41 +105,65 @@ class MinusLaplacian
...
   using SubdomainOperator = elliptic::dg::subdomain_operator::SubdomainOperator<
       poisson_system, OptionsGroup, tmpl::list<>,
       tmpl::list<Poisson::BoundaryConditions::Robin<Dim>>>;
   using SubdomainData = ::LinearSolver::Schwarz::ElementCenteredSubdomainData<
       Dim, tmpl::list<Poisson::Tags::Field>>;
+  // Associates Dirichlet or Neumann conditions to every external block
+  // boundary. For every configuration of this type we'll need a separate
+  // solver, which may cache the Poisson operator matrix that imposes the
+  // corresponding boundary conditions. This type is used as key in other maps,
+  // so it must be hashable. `unordered_map` isn't hashable, but `map` is.
```
Comment (wthrowe) [inline src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp]: Is this a Boost thing?  I don't see anything relevant about `std::hash`.  Just be explicit about that in the comment if so.

File: src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp [hunk 2]
```diff
@@ -47,18 +64,27 @@ struct MinusLaplacian {
...
+ * conditions on the Poisson sub-problems. For tensor components and element
+ * faces where the original boundary conditions are of Dirichlet-type we choose
+ * homogeneous Dirichlet boundary conditions, and for those where the original
+ * boundary conditions are of Neumann-type we choose homogeneous Neumann
+ * boundary conditions. This means we may end up with more than one distinct
+ * Poisson operator on subdomains with external boundaries, one per unique
+ * combination of element face and boundary-condition type among the tensor
+ * components.
  *
- * \tparam Dim Spatial dimension
```
Comment (wthrowe) [inline src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:61]: Why did you remove this?

**Review by wthrowe — CHANGES_REQUESTED** [review]
You can squash any changes directly.

File: src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp]: Yes, it's a Boost thing. `boost:hash` exists for `std::map`, but doesn't for `std::unordered_map`. I edited this comment.

File: src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp (hunk 2 above)
Comment (nilsvu) [reply src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp:61]: restored 👍

**Review by wthrowe — APPROVED** [review]
Looks good.

**Review by nikwit — APPROVED** [review]
LGTM

**Comment by nilsvu** [comment]
Thanks for reviewing this one @wthrowe @nikwit !
